[Dorcas Tan] Duke Increments

dorcastan opened the pull request.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

dorcastan committed:
Merge pull request #1 from nus-cs2103-AY1920S1/master

update project base before starting

dorcastan committed:
Merge pull request #2 from nus-cs2103-AY1920S1/master

update fixed tutorial

j-lum committed:
Change file mode on `gradle` to be executable

j-lum committed:
Merge pull request #12 from j-lum/gradle+x

Change file mode on `gradle` to be executable (#9)

dorcastan committed:
Add: Skeleton - greet, echo and exit functions

dorcastan committed:
Add: skeleton - basic greet, echo and exit functions

dorcastan committed:
Add: Ability to store user input before displaying

Text input from user is stored in a list (for now) and can be displayed
upon request using the command "list".

dorcastan committed:
Merge branch 'master' of https://github.com/dorcastan/duke

dorcastan committed:
Fix: documentation for Duke main class

dorcastan committed:
Add: Ability to mark tasks as done or undone

j-lum committed:
Add configuration for console applications

Gradle defaults to an empty stdin which results in runtime exceptions
when attempting to read from `System.in`. Let's add some sensible
defaults for students who may still need to work with the standard
input stream.

dorcastan committed:
Add: Different task types - Todo, Event, Deadline

User input (apart from the command "bye") must contain at least 4
characters and must follow the required format precisely.

dorcastan committed:
Fix: Change indentation for Duke's messages

Increases indentation for Duke's output text from 4 to 5 spaces to
match project specification.

dorcastan committed:
Add: Tests for 6 basic commands using text UI

dorcastan committed:
Fix: IOOBE - Change string method used

Changes substring(...) to startsWith(...); fixes
IndexOutOfBoundsException but input is no longer case insensitive.
Added skeleton for handling exceptions.

j-lum committed:
Merge pull request #13 from j-lum/javaexec

Add configuration for console applications

dorcastan committed:
Add: Error handling - only 7 commands allowed

Requires input to be of the form "bye", "list", "done [taskIndex]",
"undo [taskIndex]", "todo [description]", "deadline [description /by
[time]", or "event [description] /at [time]".

dorcastan committed:
Fix: Access modifiers and method order.

Minor changes to clean up the code.

dorcastan committed:
Add: Delete function

Users can remove tasks from the list using "delete [taskIndex]".
Tests have not been updated.

dorcastan committed:
Fix: Access modifiers and final variable names.

dorcastan committed:
Fix: Minor format issues (comments, line breaks)

dorcastan committed:
Refactor: Hide app logic in class DukeLogic

dorcastan committed:
Add: Basic save functionality

Duke automatically saves to [root]/data/duke.txt whenever the task
list is modified. Output is formatted using toString().
No file reading functionality yet.

dorcastan committed:
Fix: Format of saved data

dorcastan committed:
Add: Read-task-from-file functionality

Duke recovers the task list from the previous session if a valid data file
is present. Data file must be properly formatted.

dorcastan committed:
Add: Date parsing for Events and Deadlines

Duke now understands a limited range of date/time formats for Event timings
and Deadline due dates. Given an invalid date format, it substitutes the
current date/time.

dorcastan committed:
Merge branch 'branch-Level-7'

Incorporate Save functionality.

dorcastan committed:
Merge branch 'branch-Level-8'

Incorporate Date and Time functionality.

dorcastan committed:
Fix: Merge issues - Date format for saving

Duke now loads hard disk data and runs correctly.

dorcastan committed:
Fix: Order of Date Formats

Previous order involved simpler patterns being parsed before more complex
ones, causing date info from duke.txt to be processed incorrectly.

dorcastan committed:
REFACTOR: Separate UI, storage, parser, Commands

Reorganised code to extract out closely related sections as classes. Minimal
changes to working code i.e. not optimised yet.

dorcastan committed:
Refactor: Packages for duke, command, task

Move Tasks and Commands into separate packages. Move Parser, Storage,
TextUi, and DukeException into duke package. Change access modifiers
accordingly.

dorcastan committed:
Merge pull request #4 from nus-cs2103-AY1920S1/gradle

Add Gradle

dorcastan committed:
Fix: Move Duke main class to package duke

To make 'gradle run' work.

dorcastan committed:
Add: Gradle plugins and dependencies

Add Checkstyle, Shadow (jar), and JUnit. Following Gradle tutorial.

dorcastan committed:
Fix: Style issues

To comply with the suggested checkstyle.

dorcastan committed:
Add: JUnit test for Parser

Part #1 of unit test requirement for A-JUnit.

dorcastan committed:
Refactor: Make Storage an interface

Improve OOP practice and enable independent testing.

dorcastan committed:
Fix: Make Task getType() method public

For ease of testing.

dorcastan committed:
Add: Unit tests for Command and DeadlineCommand

Test all key methods in classes. Use StorageStub to (partially) isolate
SUT.

dorcastan committed:
Fix: .gitignore update

dorcastan committed:
Fix: pre-release touch ups

Change build version. Allow Duke to automatically create a data log
file if it does not exist. Minor edit to UI welcome message.

dorcastan committed:
Add: Javadoc for main classes

Minimal Javadoc comments in certain places.

dorcastan committed:
Add: Find functionality

Enables basic searching of task list using any keyword contained
in the full task description.

dorcastan committed:
Merge branch 'branch-Level-9'

Incorporate javadoc comments and find functionality

dorcastan committed:
Fix: Spacing for existing javadoc comments

Make minor tweaks to code style to satisfy A-CodingStandard
requirements. Code already passes Gradle checkstyle.

dorcastan committed:
Merge branch 'branch-A-CodingStandard'

Keeps most javadoc comments from master as they were previously merged from
A-JavaDoc.

dorcastan committed:
Fix: JavaDoc comment style and access modifiers

Added period '.' to certain one-line comments to comply with
checkstyle. Changed package-private access modifiers to public or
protected to better reflect access settings.

dorcastan committed:
Add: JavaFX "Hello World" program

Set up Duke as a JavaFX application and create Launcher class
according to JavaFX Tutorial Part 1.

dorcastan committed:
Fix: JavaFX application font

dorcastan committed:
Add: GUI layout skeleton

Set elements in position for Duke GUI, according to JavaFX Tutorial
Part 2.

dorcastan committed:
Add: User interaction via GUI

Allow users to interact with Duke logic using GUI, following steps
from JavaFX Tutorial Part 3. User and Duke dialog boxes are displayed
in typical chat format on opposite sides of the screen. Details of
Duke's responses still need to be ironed out.

Not supported: Save functionality.

dorcastan committed:
Add: Basic FXML and separate Main class

Following JavaFX Tutorial Part 4, use FXML to specify most of the
existing GUI features. Create a Main class and move GUI-related
code from Duke to MainWindow class.

dorcastan committed:
Revert "A-Checkstyle"

This reverts commit b3f3be55aae9cf759fb6e80920dfe5b32b735010.
Accidentally merged branch-Level-10 to master and pushed to origin.
More GUI improvements to be made before merging.

gabrielchao commented:
General comments:
- Good use of interfaces for Storage and HardDiskStorage
- Comprehensive, detailed, and informative JavaDocs on every method and field of every class
- Tight adherence to coding standards

gabrielchao commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,117 @@
+ */
+public abstract class Command {
+    /** Details of the command. */
+    protected String details;
```
Perhaps the meaning of "details" could be better documented? I had to check the uses of this field in the concrete class to verify that it refers to the arguments of the Command.

gabrielchao commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,136 @@
+     */
+    private String getStatusIcon() {
+        return isDone ? "+" : " ";
+        // return isDone ? "\u2713" : "\u2718"; // return tick or X symbols
```
If I'm not wrong, outputting to a `Printstream` with UTF-8 `Charset` allows for Unicode to be printed.

gabrielchao commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,136 @@
+     * @return Date     Date corresponding to the given date String, or the
+     *                  current instant
+     */
+    static Date parseDate(String date) {
```
This is subjective, but would it be a better implementation of the Single Responsibility Principle to have the static parse methods be part of the `Parser` class instead? Or maybe even have them in a `DateParser` class?

dorcastan committed:
Fix: Javadoc comment style

Minor changes (e.g. add period after parameters) to comply with the
CS2103T coding standard.

gabrielchao commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,46 @@
+                foundTasks.add(task);
+            }
+        }
+        ui.showList(foundTasks);
```
Would it be more ideal to print the found tasks with their original list numbers? Then there would be no confusion if the user wanted to perform commands on the found tasks.

gabrielchao reviewed (commented):
This code was pleasant to review. Well done!

keiteo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,107 @@
+package duke;
```
Can be packaged into smaller parts for easier management (e.g. duke.util.Parser).

keiteo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,107 @@
+        String command = input.strip();
+        validate(command);
+
+        if (command.startsWith("done")) {
```
Consider using switch cases if there are too many if else statements!

keiteo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    /**
+     * Throws an exception if the given input does not have a valid format.
```
This could be the description for @throws DukeException part. A better way to describe this method would be something like "Validates a command" or "Checks for valid format in a command" etc.

keiteo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,107 @@
+     */
+    private static void validate(String input) throws DukeException {
+        if (input.startsWith("done") || input.startsWith("undo")) {
+            if (input.length() < 6) {
```
Try to avoid nested if statements to make code a little bit cleaner.

keiteo commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,33 @@
+     * Constructs a bye Command, which sets the boolean value isExit to true.
+     */
+    public ByeCommand() {
+        super("");
```
Maybe you can overload the Command constructor without parameters instead of passing in an empty string.

keiteo commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,33 @@
+     */
+    @Override
+    public String execute(TaskList tasks, TextUi ui, Storage storage) {
+        String textToDisplay = "Bye. Hope to see you again soon!";
```
Maybe the textToDisplay can be abstracted to the UI class it's more related to UI components?

keiteo commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,118 @@
+    Command(String details) {
+        this.details = details;
+        this.isExit = false;
+        this.isDone = false;
```
isExit and isDone values can be assigned in the attribute field instead of the constructor since it's kinda hardcoded. e.g. just directly declare `protected boolean isDone = false;`

keiteo commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,33 @@
+     */
+    public ByeCommand() {
+        super("");
+        this.isExit = true;
```
Try to change parent class attributes with a setter or an overloaded constructor. Direct overwriting from the child class kinda defeats the encapsulation aspect of OOP.

keiteo commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,46 @@
+        TaskList foundTasks = new TaskList();
+        for (int i = 0; i < tasks.size(); i++) {
+            Task task = tasks.get(i);
+            if (task.toString().contains(this.details)) {
```
Use a getter instead of `this.details` for slightly better OOP.

keiteo commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,38 @@
+    public String execute(TaskList tasks, TextUi ui, Storage storage) {
+        String response;
+        if (tasks.isEmpty()) {
+            response = "You have no tasks now. Hooray!";
```
Can make your code cleaner using a single if statement e.g
```
if (task.isEmpty()) {
return "you have no task";
}
// do your thing
return response;
```

keiteo reviewed (commented):
Code is generally clean and easy to read. Coding style is strictly adhered, excellent job!

Some tips:
- switch cases would be good if your if else statements get too big
- package parser, storage class etc. for better organisation
- try to keep encapsulation in mind when writing parent child classes

dorcastan committed:
Fix: Style issues (.* import, javadocs)

Reorder @return and @throws statements in CommandTest and replace
wildcard import in CommandTest to pass Gradle checkstyle test.

dorcastan committed:
Merge branch 'branch-Level-10'

dorcastan committed:
Refactor: Parser class to increase readability

Make use of 'switch' statements, separate different types of validation,
move command-specific checks (i.e. "/at" and "/by") into respective
Command classes. Changes made in response to PR comments.

Also: Update version in build.gradle to v0.1.5.

dorcastan committed:
Branch A-Code-Quality (#7)

* Add: Util and UI packages to improve organisation

Too many classes in main package duke. Parser, Storage, and TextUI
classes were moved into duke.util, and JavaFX classes into duke.ui.

Also fixes JavaDoc format for block tags in the classes that were
modified.

* Fix: Wrong path to MainWindow class

Path was not updated after previous refactor.

* Refactor: Date parsing and status icon checking

Refactor Task class for better OOP.
1. Move Date parsing functionality from Task into a separate DateParser.
2. Allow Task status icons to be changed without affecting other classes
such as HardDiskStorage.

* Fix: Wrong class paths in Test classes

Test classes were not updated after refactoring.
Update import statements for Storage and TextUi classes, and move
ParserTest into duke.util package.

* Fix: Application main class in build.gradle

Main class name of application was not updated after GUI was made.
Jar archives created using Gradle ShadowJar can now be run properly.

* Refactor: Command class constructors

Added more overloaded constructors in Command class, to eliminate calls
to super("") (empty string) from child classes.

* Add: Assertions and miscellaneous refactoring

Add assertions in key methods of various classes to check that the
internal logic is correct.

Miscellaneous improvements to code quality, including changing multiple
"if" blocks to switch statements.

* Refactor: toString() method to improve readability

Made use of Java Streams instead of a "for" loop to shorten the code.

* Fix: Minor style issues (variable names)

Replaced abbreviations with more descriptive names.

dorcastan committed:
Add: Priority levels for Tasks

Allows users to mark Tasks with one out of four possible Priority
levels: High, Medium, Low, or None. Tasks with priority higher than
None are displayed with their priority level before task description.

Alternative keywords for priority levels Medium (e.g. mid, quite) and
None (no) are provided. Priority levels are NOT saved after exiting
Duke.

dorcastan committed:
Fix: Minor JavaDocs content

Add description of Priority enumeration, and other minor updates.

dorcastan committed:
Update version to v0.1.6 (build.gradle)

dorcastan committed:
Update version to v0.1.7 (build.gradle)

In preparation for next round of changes (after Priority).

dorcastan committed:
Fix: Problems with Save functionality in GUI

Duke has been unable to create the data file /data/duke.txt since
migrating from CLI to GUI. Issue was a missing relative file path.
Resolved by using System property "user.dir" when creating file path
in HardDiskStorage.

Also added load-tasks-from-disk functionality to the Duke constructor
used by JavaFX.

dorcastan committed:
Fix: Show found tasks with original index

A FindCommand made a new list for found tasks, so index numbers shown
did not correspond to the numbers required for operations. Issue was
highlighted in peer review. Found Tasks are now displayed with their
original index numbers.

dorcastan committed:
Add: Case insensitive command types

Command types (first word) can now be parsed regardless of what case
they are in.

dorcastan committed:
Fix: Date parsing and formatting

Date parsing previously involved Date and SimpleDateFormat, which have
now been replaced by the newer Java classes LocalDateTime and
DateTimeFormatter.

To avoid having to change Task classes whenever DateParser is modified,
toString methods of time-related Tasks use a default date/time format
given by DateParser.

dorcastan committed:
Fix: TODOs

Remove unnecessary or completed TODO comments.

dorcastan committed:
Fix: Truncation of Duke response

Remove maximum height for the dialog container so that Duke responses
(e.g. lists of tasks) can be viewed in their entirety.

dorcastan committed:
Add: Welcome message for Snowball

Rebrands Duke as Snowball. Add welcome message when application is
launched. Create SpeechMaker class to handle Snowball speech.

dorcastan committed:
Add: Styling via FXML

Change user display picture. Add background and fix padding for dialog
box text.

dorcastan committed:
Fix: Extend application window height

So that more lines can be seen at a time.

dorcastan committed:
Fix: Better use of SpeechMaker

Move more of Snowball's responses to the SpeechMaker class for better
OOP. Remove remaining references to Duke personality, along with unused
code. (The only use of Duke should be "the Duke application".)

dorcastan committed:
Add: Product name to title bar

Show product title in window title bar for clarity.

dorcastan committed:
Add: Save functionality for Priority

Implement storage and parsing of task Priority levels. Also, rename
TaskList indexed-string method as toIndexedString() for consistency with
toString() methods.

dorcastan committed:
Refactor: Task constructor to include Priority

Improve the loading of existing tasks from storage by using overloaded
constructors to create new Tasks with differing Priority levels.

Minor Javadocs formatting changes.

dorcastan committed:
Fix: Checkstyle issues

Settle missing periods in Javadocs, indentation, etc. to pass Gradle
checkstyle.

dorcastan committed:
Add: User Guide

Describe what Snowball is and how to use it. Add explanations for all
commands and command parameter restrictions. Include some examples, but
more examples can be shown especially for Add commands.

dorcastan committed:
Fix: Minor format issues (date parsing & display)

Add a shorter format for inputting dates conveniently. Make task lists
appear more presentable in Snowball's responses.

dorcastan committed:
Set theme jekyll-theme-modernist

dorcastan committed:
Add: User Guide in docs folder

User Guide was mistakenly added to source README.md in a previous
commit.

dorcastan committed:
Set theme jekyll-theme-merlot

dorcastan committed:
Set theme jekyll-theme-merlot

dorcastan committed:
Fix: Revert README.md to original tutorial

User guide was mistakenly published to the wrong README.md.

dorcastan committed:
Merge branch 'master' of https://github.com/dorcastan/duke

Origin was updated to publish GitHub Pages

dorcastan committed:
Add: Product screenshot

Show user interface and selected commands.

dorcastan committed:
Set theme jekyll-theme-merlot

dorcastan committed:
Update to v0.2.0 (build.gradle)

dorcastan committed:
Merge branch 'master' of https://github.com/dorcastan/duke

dorcastan committed:
Fix: Cross-platform functionality for .jar

Apply suggested modifications to build.gradle so that Snowball can work
across Mac, Windows and Linux distributions.

Update from v0.2.0 to v0.2.1

dorcastan committed:
Fix: Path to Snowball image (User Guide)

Replaced relative path with URL for correct displaying on GitHub Pages.

dorcastan committed:
Add: Example usages to user guide

Improve documentation of command usage to match the recommended
format (example usage + expected outcome).

dorcastan committed:
Fix: Exit functionality

Close the application when a bye message is shown, and update User
Guide accordingly.
Also changed naming of standard Snowball messages to comply with code
style standard for static final attributes.

dorcastan committed:
Refactor: String responses from Snowball

Move all text responses from individual Command classes to the
SpeechMaker class for better separation of responsibilities.

dorcastan committed:
Fix: Javadocs for SpeechMaker

Add more descriptive comments for messages.
